src/vea/vea_alloc.c

```diff
@@ -328,3 +502,151 @@ persistent_alloc(struct vea_space_info *vsi, struct vea_free_extent *vfe)
+	if (!isclr_range((uint8_t *)bitmap->vfb_bitmaps,
+			 bit_at, bit_at + bits_nr - 1)) {
+		D_ERROR("bitmap already set in the range.["DF_U64", %u]\n",
+			 blk_off, blk_cnt);
+		return -DER_INVAL;
+	}
+
+	if (vsi_umem) {
+		/* Todo this could be optimized */
+		rc = umem_tx_add_ptr(vsi_umem, bitmap->vfb_bitmaps,
+				     bitmap->vfb_bitmap_sz << 3);
+		if (rc)
+			return rc;
+	}
+
+	D_ASSERT(bit_at + bits_nr <= bitmap->vfb_bitmap_sz * 64);
+	setbits64(bitmap->vfb_bitmaps, bit_at, bits_nr);
+
+	return 0;
+}
+
+static int
+persistent_alloc_bitmap(struct vea_space_info *vsi, struct vea_free_extent *vfe)
+{
+	struct vea_free_bitmap	*found;
+	daos_handle_t		 btr_hdl;
+	d_iov_t			 key_in, val;
+	int			 rc, opc = BTR_PROBE_LE;
+
+	D_ASSERT(umem_tx_inprogress(vsi->vsi_umem) ||
+		 vsi->vsi_umem->umm_id == UMEM_CLASS_VMEM);
+	D_ASSERT(vfe->vfe_blk_off != VEA_HINT_OFF_INVAL);
+	D_ASSERT(vfe->vfe_blk_cnt > 0);
+
+	btr_hdl = vsi->vsi_md_bitmap_btr;
+	D_ASSERT(daos_handle_is_valid(btr_hdl));
+
+	D_DEBUG(DB_IO, "Persistent alloc ["DF_U64", %u]\n",
+		vfe->vfe_blk_off, vfe->vfe_blk_cnt);
+
+	/* Fetch & operate on the in-tree record */
+	d_iov_set(&key_in, &vfe->vfe_blk_off, sizeof(vfe->vfe_blk_off));
+	d_iov_set(&val, NULL, sizeof(*found));
+
+	rc = dbtree_fetch(btr_hdl, opc, DAOS_INTENT_DEFAULT, &key_in, NULL,
+			  &val);
+	if (rc) {
+		D_ERROR("failed to find extent ["DF_U64", %u]\n",
+			vfe->vfe_blk_off, vfe->vfe_blk_cnt);
+		return rc;
+	}
+
+	found = (struct vea_free_bitmap *)val.iov_buf;
+	rc = verify_bitmap_entry(found);
+	if (rc) {
+		D_ERROR("bitmap corrupted ["DF_U64", %u]\n",
+			found->vfb_blk_off, found->vfb_class);
+		return rc;
+	}
+
+	return bitmap_set_range(vsi->vsi_umem, found,
+				vfe->vfe_blk_off, vfe->vfe_blk_cnt);
+}
+
+int
+persistent_alloc(struct vea_space_info *vsi, struct vea_free_entry *vfe, void *private)
+{
+	struct vea_bitmap_entry	*bitmap_entry = (struct vea_bitmap_entry *)private;
+
+	if (vfe->vfe_type == VEA_FREE_ENTRY_EXTENT)
+		return persistent_alloc_extent(vsi, &vfe->vfe_ext);
+
+	D_ASSERT(bitmap_entry != NULL);
+	/* if this bitmap is new */
+	if (bitmap_entry->vbe_is_new) {
+		d_iov_t			 key, val;
+		struct vea_free_bitmap	*bitmap;
+		int			 rc;
+		struct vea_free_extent   extent;
+		daos_handle_t		 btr_hdl = vsi->vsi_md_bitmap_btr;
+
+		extent = vfe->vfe_ext;
+		extent.vfe_blk_off = bitmap_entry->vbe_bitmap.vfb_blk_off;
+		extent.vfe_blk_cnt = bitmap_entry->vbe_bitmap.vfb_blk_cnt;
+		rc = persistent_alloc_extent(vsi, &extent);
+		if (rc)
+			return rc;
+
+		D_ALLOC(bitmap, alloc_free_bitmap_size(bitmap_entry->vbe_bitmap.vfb_bitmap_sz));
+		if (!bitmap)
+			return -DER_NOMEM;
+
+		D_ASSERT(vfe->vfe_ext.vfe_blk_cnt != 0);
+		bitmap->vfb_blk_off = extent.vfe_blk_off;
+		bitmap->vfb_class = vfe->vfe_ext.vfe_blk_cnt;
+		bitmap->vfb_blk_cnt = bitmap_entry->vbe_bitmap.vfb_blk_cnt;
+		bitmap->vfb_bitmap_sz = bitmap_entry->vbe_bitmap.vfb_bitmap_sz;
+		rc = bitmap_set_range(NULL, bitmap, vfe->vfe_ext.vfe_blk_off,
```

Comment:
The umem_tx_add_ptr() in bitmap_set_range() is mistakenly skipped, that could cause problem if the transaction created bitmap chunk failed/aborted. (The bitmap is updated directly without undo log).

Reply:
Sure.